### PR TITLE
Added pastebin example

### PR DESCRIPTION
**Pull request by 414owen**
This example shows frontend routing and redirection

### CODE REVIEW

**Comment by 3noch** [comment]
What do you think this example adds that the url-shortener example lacks?

**Comment by 414owen** [comment]
Mainly frontend routing and redirection

**Comment by 3noch** [comment]
Is there a way you could achieve both things in the same example do you think? It would cut down on the "duplicateness" of these examples.

**Comment by 414owen** [comment]
Well it would be hard to incorporate these things into the url-shortener, but to be honest the url-shortener doesn't show off any features that this example doesn't, so we could just close it in favour of this.

**Comment by 3noch** [comment]
I like that idea

File: pastebin/common/common.cabal [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+cabal-version: >= 1.2
+build-type: Simple
+
+library
+  hs-source-dirs: src
+  build-depends: base
+               , obelisk-route
+               , mtl
+               , text
+  default-extensions:
```
Comment (alexfmpe) [inline pastebin/common/common.cabal:12]: Why these two in particular?

File: pastebin/common/src/Common/Api.hs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+module Common.Api where
+
+import Data.Aeson
+import Data.Text (Text)
+import GHC.Generics (Generic)
+
+newtype PasteRequest = PasteRequest Text deriving (Generic, ToJSON, FromJSON)
```
Comment (alexfmpe) [inline pastebin/common/src/Common/Api.hs]: Should we specify deriving strategy for the serialization classes? It makes a difference whether it’s GND or not - I think changing between those two accidentally would be a breaking change

File: pastebin/common/src/Common/Route.hs [hunk 3]
```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (alexfmpe) [inline pastebin/common/src/Common/Route.hs:8]: Maybe alphabetize these?

File: pastebin/common/src/Common/Api.hs (hunk 2 above)
Comment (alexfmpe) [inline pastebin/common/src/Common/Api.hs]: Missing the extractor, if it matters

File: pastebin/frontend/src/Frontend.hs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  , MonadFix m
+  , MonadHold t m
+  , TriggerEvent t m
+  , PostBuild t m
+  , PerformEvent t m
+  , Prerender js t m
+  )
+
+type FrontendWidget js t m =
+  ( AppWidget js t m
```
Comment (alexfmpe) [inline pastebin/frontend/src/Frontend.hs:38]: I find these names confusing. What do AppWidget and FrontendWidget mean? Doesn’t look like I could use either of them in a browser with javascript disabled, even though that is a frontend.

File: pastebin/frontend/src/Frontend.hs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  Just t -> el "pre" $ text t
+
+getPasteUrl :: Id Paste -> Text
+getPasteUrl pasteId = renderBackendRoute checkedFullRouteEncoder $ BackendRoute_GetPaste :/ pasteId
+
+viewClientSide :: FrontendWidget js t m => Dynamic t (Id Paste) -> m ()
+viewClientSide pasteId = do
+  onload <- getPostBuild
+  initialPasteLoad <- getAndDecode $ getPasteUrl <$> tag (current pasteId) onload
+  res <- holdDyn viewPlaceholder (showPasteResult <$> initialPasteLoad)
```
Comment (alexfmpe) [inline pastebin/frontend/src/Frontend.hs]: `widgetHold_` seems like a better fit

File: pastebin/frontend/src/Frontend.hs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+app :: (AppWidget js t m, SetRoute t (R FrontendRoute) m) => RoutedT t (R FrontendRoute) m ()
+app =
+  subRoute_ $ \case
+    FrontendRoute_Main -> do
+      state <- pasteInput
+      setRoute $ (FrontendRoute_ViewPaste :/) . Id <$> fmapMaybe id (stateToMaybeKey <$> updated state)
+    FrontendRoute_ViewPaste -> do
+      pasteId <- askRoute
+      _ <- prerender
```
Comment (alexfmpe) [inline pastebin/frontend/src/Frontend.hs]: `prerender_`

File: pastebin/common/common.cabal (hunk 1 above)
Comment (3noch) [reply pastebin/common/common.cabal:12]: This is from the skeleton...pretty silly though.

File: pastebin/common/src/Common/Api.hs (hunk 2 above)
Comment (3noch) [reply pastebin/common/src/Common/Api.hs]: Breaking in the sense that if this were a public API. This is not a public API though so as long as the frontend and backend agree it's fine. However, using deriving strategies seems like a good idea in general to me.

@414owen I.e. `deriving stock Generic deriving newtype (ToJSON, FromJSON)`

File: pastebin/common/src/Common/Route.hs (hunk 3 above)
Comment (3noch) [reply pastebin/common/src/Common/Route.hs:8]: Good catch. Yes we try to do that.

File: pastebin/common/src/Common/Api.hs (hunk 2 above)
Comment (alexfmpe) [reply pastebin/common/src/Common/Api.hs]: I was thinking more along the lines of storing to any persistent format and failing to read it at a later date.

File: pastebin/common/common.cabal (hunk 1 above)
Comment (alexfmpe) [reply pastebin/common/common.cabal:12]: Took this upstream

File: pastebin/common/src/Common/Api.hs (hunk 2 above)
Comment (ApolloUnicorn) [reply pastebin/common/src/Common/Api.hs]: @414owen did you see above comment?

Comment (414owen) [reply pastebin/common/src/Common/Api.hs]: Ah yeah, adding now

**Comment by 3noch** [comment]
Last but not least, we need to bump this to latest obelisk master.

**Pull request merged**